sips/sip-012/sip-012-cost-limits-network-upgrade.md

```diff
@@ -0,0 +1,742 @@
+
+License: BSD 2-Clause
+
+Sign-off:
+
+Discussions-To: https://github.com/stacksgov/sips
+
+# Abstract
+
+The current Clarity cost limits were set very conservatively in Stacks 2.0:
+transactions with contract-calls frequently exceed these limits, which
+negatively affects transaction throughput. This SIP proposes an update to these
+cost-limits via a network upgrade and further, that the network upgrade be
+executed at a block height chosen by an off-chain process described in this SIP.
+
+
+# License and Copyright
+
+This SIP is made available under the terms of the Creative Commons CC0 1.0
+Universal license, available at
+https://creativecommons.org/publicdomain/zero/1.0/ This SIP’s copyright is held
+by the Stacks Open Internet Foundation.
+
+
+# Introduction
+
+Blocks on the Stacks blockchain tend to have anywhere between 10 to 50
+transactions per block -- this is lower than the demands of many workloads and
+also lower than the theoretical maximum one would expect. On the other hand, the
+mempool consistently has hundreds of valid transactions pending at any given
+time; at peak there have been several thousand pending transactions. So what is
+preventing more transactions from being included in blocks?
+
+The answer is cost-limits. An analysis of “full blocks” (meaning blocks where at
+least one cost dimension exceeds the block limit -- see SIP-006 for a
+description of all cost dimensions) found that roughly 72% of those blocks hit
+the `runtime` limit and ~27% hit the `read_count` limit. Another analysis of all
+transactions rejected due to cost limits found that 90% of those transactions
+exceeded the `runtime` limit. So the `runtime` limits are the primary bottleneck
+right now.
+
+In the last few months, the
+[clarity-benchmarking](https://github.com/blockstack/clarity-benchmarking)
+project has done rigorous benchmarking on contemporary hardware to come up with
+more accurate cost-limits, with a focus on the `runtime` limits. The updated
+cost-limits are described in detail in [this forum
+post](https://forum.stacks.org/t/more-accurate-cost-functions-for-clarity-native-functions/12386).
+
+Any modification of cost-limits is a consensus-breaking change. We believe that
+there is broad community support for changing the cost-limits, the question is exactly
+how and when they go into effect. A previous proposal suggested using a voting
+contract to determine the block height at which a network-upgrade, described in
+detail in [this Github
+discussion](https://github.com/blockstack/stacks-blockchain/discussions/2845).
+Unfortunately, this path would take at least 4 months in the best-case scenario.
+
+This SIP posits that the ongoing network congestion warrants a more expedient
+route to change the cost-limits, one that does not rely on an on-chain voting
+contract. Equally of note, we consider this SIP as a method of last resort but
+that considering the circumstances an exception is justified. All future network
```

Comment:
"All future network upgrade..." This was true already before SIP-012 proposal. This SIP could define a list of circumstances when a similar process of SIP-012 could be applied. 

To me, it is not clear whether the sip defines a general process to be applied under exceptional circumstances (how to set the block height, etc) or whether the SIP defines a one-off process (because applying the same values of Appendix A would not change anything after the values have be used as new cost limits).
 
To avoid confusion, we should only use "network upgrade", or introduce `hard-fork` as an equivalent term. (Maybe there is already in one of the sips).

Reply:
SIP-011 is an attempt to clarify when breaking changes in the consensus layer are allowed to occur, but it's still a draft.

I'm of the opinion that this SIP should only define a one-off process, because there's already a general process being developed.  There can be a subsequent SIP that re-proposes this process as a general process if it turns out to work really well.